### ISSUE TITLE
overloads not next to each other result in confusing errors

### ISSUE THREAD

**Issue by thomasballinger** [python/mypy#1614]
Overloads not next to each other are correctly rejected, but with confusing error (related to, maybe the same as https://github.com/python/mypy/issues/409)

~~~
from typing import overload
@overload
def abs(x: int): pass
1
@overload
def abs(x: float): pass


tmpstub.pyi:3: error: "object" not callable
tmpstub.pyi:6: error: Name 'abs' already defined
tmpstub.pyi:6: error: "object" not callable
~~~

**Comment by gvanrossum** [comment]
Wow. That's crazy! If you want to debug this further yourself, try putting a PDB trap (`import pdb; pdb.set_trace()` in Errors.report() in mypy/errors.py, and crawl back from the first error.

**Comment by thomasballinger** [comment]
Currently, `OverloadedFuncDef` instances are created by mypy.parser.Parser.try_combine_overload (https://github.com/python/mypy/blob/57359c13da488f429e74e2c425be21f5d98af339/mypy/parse.py#L880) whenever decorated functions with the same name are consecutive in a file, regardless of which decorators are used. Once functions combined into an `OverloadedFuncDef` they have better error messages down the pipeline to enforce those decorators being `overload`.

In order to get better error messages for this situation, it would be nice to classify any function used with `@overload` as an `OverloadedFuncDef` so checks could happen in that class, instead of combining functions any time they have the same name and are consecutive. I don't understand why it might be important for combining of functions to be so general.

A messier alternative would be to look for `@overload` when processing decorators and report that it's been misused somehow.

**Comment by gvanrossum** [comment]
Probably best to check with @JukkaL, as I don't know the history of this code.

**Comment by thomasballinger** [comment]
here's a draft PR that does with code what I describe above: https://github.com/python/mypy/pull/1638

**Comment by gvanrossum** [comment]
Fixed by 3fe364f2a118b4b6ba286d5790a968be94a81abe.
